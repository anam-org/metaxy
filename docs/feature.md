## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
