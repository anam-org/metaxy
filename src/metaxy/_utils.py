import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
