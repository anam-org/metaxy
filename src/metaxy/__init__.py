from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
