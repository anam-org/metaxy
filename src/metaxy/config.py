"""Configuration system for Metaxy using pydantic-settings."""

from pathlib import Path
from typing import TYPE_CHECKING, Any, TypeVar, overload

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Fallback for Python 3.10

from pydantic import Field as PydanticField
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
)

if TYPE_CHECKING:
    from metaxy.metadata_store.base import MetadataStore
    from metaxy.models.feature import FeatureRegistry

T = TypeVar("T")


def load_registry(registry_path: str) -> "FeatureRegistry":
    """Load registry from import path.

    Args:
        registry_path: Import path like "myapp.features:my_registry"

    Returns:
        FeatureRegistry instance

    Raises:
        ValueError: If path is invalid
        ImportError: If module or registry not found
        TypeError: If loaded object is not a FeatureRegistry
    """
    from metaxy.models.feature import FeatureRegistry

    if ":" not in registry_path:
        raise ValueError(
            f"Invalid registry path: {registry_path}. "
            f"Expected format: 'module.path:registry_name'"
        )

    module_path, registry_name = registry_path.split(":", 1)

    try:
        module = __import__(module_path, fromlist=[registry_name])
        registry = getattr(module, registry_name)
    except (ImportError, AttributeError) as e:
        raise ImportError(f"Failed to load registry from {registry_path}: {e}") from e

    if not isinstance(registry, FeatureRegistry):
        raise TypeError(
            f"Registry at {registry_path} is not a FeatureRegistry instance. "
            f"Got: {type(registry)}"
        )

    return registry


class TomlConfigSettingsSource(PydanticBaseSettingsSource):
    """Custom settings source for TOML configuration files.

    Auto-discovers configuration in this order:
    1. Explicit file path if provided
    2. metaxy.toml in current directory (preferred)
    3. pyproject.toml [tool.metaxy] section (fallback)
    4. No config (returns empty dict)
    """

    def __init__(self, settings_cls: type[BaseSettings], toml_file: Path | None = None):
        super().__init__(settings_cls)
        self.toml_file = toml_file or self._discover_config_file()
        self.toml_data = self._load_toml()

    def _discover_config_file(self) -> Path | None:
        """Auto-discover config file."""
        # Prefer metaxy.toml
        if Path("metaxy.toml").exists():
            return Path("metaxy.toml")

        # Fallback to pyproject.toml
        if Path("pyproject.toml").exists():
            return Path("pyproject.toml")

        return None

    def _load_toml(self) -> dict[str, Any]:
        """Load TOML file and extract metaxy config."""
        if self.toml_file is None:
            return {}

        with open(self.toml_file, "rb") as f:
            data = tomllib.load(f)

        # Extract [tool.metaxy] from pyproject.toml or root from metaxy.toml
        if self.toml_file.name == "pyproject.toml":
            return data.get("tool", {}).get("metaxy", {})
        else:
            return data

    def get_field_value(self, field: Any, field_name: str) -> tuple[Any, str, bool]:
        """Get field value from TOML data."""
        field_value = self.toml_data.get(field_name)
        return field_value, field_name, False

    def __call__(self) -> dict[str, Any]:
        """Return all settings from TOML."""
        return self.toml_data


class StoreConfig(BaseSettings):
    """Configuration for a single metadata store.

    Structure:
        type: Full import path to store class
        config: Dict of all configuration (including fallback_stores)

    Example:
        >>> config = StoreConfig(
        ...     type="metaxy_delta.DeltaMetadataStore",
        ...     config={
        ...         "table_uri": "s3://bucket/metadata",
        ...         "region": "us-west-2",
        ...         "fallback_stores": ["prod"],
        ...     }
        ... )
    """

    model_config = SettingsConfigDict(
        extra="forbid",  # Only type and config fields allowed
    )

    # Store class (full import path)
    type: str

    # Store configuration (all kwargs for __init__)
    # This includes fallback_stores, table_uri, db_path, storage_options, etc.
    config: dict[str, Any] = PydanticField(default_factory=dict)


class MetaxyConfig(BaseSettings):
    """Main Metaxy configuration.

    Loads from:
    1. TOML file (metaxy.toml or pyproject.toml [tool.metaxy])
    2. Environment variables (METAXY_*)
    3. Init arguments

    Priority: init > env vars > TOML

    Example:
        >>> # Auto-discover config
        >>> config = MetaxyConfig.load()
        >>>
        >>> # Get store instance
        >>> store = config.get_store("prod")
        >>>
        >>> # Override via env var
        >>> # METAXY_STORE=staging METAXY_REGISTRY=myapp.features:my_registry
        >>> config = MetaxyConfig.load()
        >>> store = config.get_store()  # Uses staging with custom registry
    """

    model_config = SettingsConfigDict(
        env_prefix="METAXY_",
        env_nested_delimiter="__",
    )

    # Store to use
    store: str = "dev"

    # Named store configurations
    stores: dict[str, StoreConfig] = PydanticField(default_factory=dict)

    # Migrations directory
    migrations_dir: str = "metaxy/migrations"

    # Entrypoints to load (list of module paths)
    entrypoints: list[str] = PydanticField(default_factory=list)

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        """Customize settings sources: init → env → TOML.

        Priority (first wins):
        1. Init arguments
        2. Environment variables
        3. TOML file
        """
        toml_settings = TomlConfigSettingsSource(settings_cls)
        return (init_settings, env_settings, toml_settings)

    @classmethod
    def load(
        cls, config_file: str | Path | None = None, *, search_parents: bool = True
    ) -> "MetaxyConfig":
        """Load config with auto-discovery and parent directory search.

        Args:
            config_file: Optional config file path (overrides auto-discovery)
            search_parents: Search parent directories for config file (default: True)

        Returns:
            Loaded config (TOML + env vars merged)

        Example:
            >>> # Auto-discover with parent search
            >>> config = MetaxyConfig.load()
            >>>
            >>> # Explicit file
            >>> config = MetaxyConfig.load("custom.toml")

            >>> # Auto-discover without parent search
            >>> config = MetaxyConfig.load(search_parents=False)
        """
        # Search for config file if not explicitly provided
        if config_file is None and search_parents:
            config_file = cls._discover_config_with_parents()

        # For explicit file, temporarily patch the TomlConfigSettingsSource
        # to use that file, then use normal instantiation
        # This ensures env vars still work

        if config_file:
            # Create a custom settings source class for this file
            toml_path = Path(config_file)

            class CustomTomlSource(TomlConfigSettingsSource):
                def __init__(self, settings_cls: type[BaseSettings]):
                    # Skip auto-discovery, use explicit file
                    super(TomlConfigSettingsSource, self).__init__(settings_cls)
                    self.toml_file = toml_path
                    self.toml_data = self._load_toml()

            # Customize sources to use custom TOML file
            original_method = cls.settings_customise_sources

            @classmethod  # type: ignore[misc]
            def custom_sources(
                cls_inner,
                settings_cls,
                init_settings,
                env_settings,
                dotenv_settings,
                file_secret_settings,
            ):
                toml_settings = CustomTomlSource(settings_cls)
                return (init_settings, env_settings, toml_settings)

            # Temporarily replace method
            cls.settings_customise_sources = custom_sources  # type: ignore[assignment]
            config = cls()
            cls.settings_customise_sources = original_method  # type: ignore[method-assign]
        else:
            # Use default sources (auto-discovery + env vars)
            config = cls()

        # Load entrypoints if configured (do this after config is created)
        if config.entrypoints:
            from metaxy.entrypoints import load_config_entrypoints

            load_config_entrypoints(config.entrypoints)

        return config

    @staticmethod
    def _discover_config_with_parents() -> Path | None:
        """Discover config file by searching current and parent directories.

        Searches for metaxy.toml or pyproject.toml in current directory,
        then iteratively searches parent directories.

        Returns:
            Path to config file if found, None otherwise
        """
        current = Path.cwd()

        while True:
            # Check for metaxy.toml (preferred)
            metaxy_toml = current / "metaxy.toml"
            if metaxy_toml.exists():
                return metaxy_toml

            # Check for pyproject.toml
            pyproject_toml = current / "pyproject.toml"
            if pyproject_toml.exists():
                return pyproject_toml

            # Move to parent
            parent = current.parent
            if parent == current:
                # Reached root
                break
            current = parent

        return None

    def get_store(
        self,
        name: str | None = None,
    ) -> "MetadataStore":
        """Instantiate metadata store by name.

        Args:
            name: Store name (uses config.store if None)

        Returns:
            Instantiated metadata store

        Raises:
            ValueError: If store name not found in config, or if fallback stores
                have different hash algorithms than the parent store
            ImportError: If store class cannot be imported

        Example:
            >>> config = MetaxyConfig.load()
            >>> store = config.get_store("prod")
            >>>
            >>> # Use default store
            >>> store = config.get_store()
        """
        from metaxy.data_versioning.hash_algorithms import HashAlgorithm

        if len(self.stores) == 0:
            raise ValueError(
                "No Metaxy stores available. They should be configured in metaxy.toml|pyproject.toml or via environment variables."
            )

        name = name or self.store

        if name not in self.stores:
            raise ValueError(
                f"Store '{name}' not found in config. "
                f"Available stores: {list(self.stores.keys())}"
            )

        store_config = self.stores[name]

        # Import store class
        store_class = self._import_class(store_config.type)

        # Extract configuration
        config_copy = store_config.config.copy()
        fallback_store_names = config_copy.pop("fallback_stores", [])

        # Get hash_algorithm from config (if specified) and convert to enum
        configured_hash_algorithm = config_copy.get("hash_algorithm")
        if configured_hash_algorithm is not None:
            # Convert string to enum if needed
            if isinstance(configured_hash_algorithm, str):
                configured_hash_algorithm = HashAlgorithm(configured_hash_algorithm)
                config_copy["hash_algorithm"] = configured_hash_algorithm
        else:
            # Use default
            configured_hash_algorithm = HashAlgorithm.XXHASH64
            config_copy["hash_algorithm"] = configured_hash_algorithm

        # Build fallback stores recursively
        fallback_stores = []
        for fallback_name in fallback_store_names:
            fallback_store = self.get_store(fallback_name)
            fallback_stores.append(fallback_store)

        # Instantiate store with config + fallback_stores
        store = store_class(
            fallback_stores=fallback_stores,
            **config_copy,
        )

        # Verify the store actually uses the hash algorithm we configured
        # (in case a store subclass overrides the default or ignores the parameter)
        if store.hash_algorithm != configured_hash_algorithm:
            raise ValueError(
                f"Store '{name}' ({store_class.__name__}) was configured with "
                f"hash_algorithm='{configured_hash_algorithm.value}' but is using "
                f"'{store.hash_algorithm.value}'. The store class may have overridden "
                f"the hash algorithm. All stores must use the same hash algorithm."
            )

        return store

    @staticmethod
    def _import_class(class_path: str) -> type:
        """Import class from module path.

        Args:
            class_path: Full import path like "metaxy.metadata_store.InMemoryMetadataStore"

        Returns:
            Imported class

        Raises:
            ImportError: If module or class not found
        """
        module_path, class_name = class_path.rsplit(".", 1)
        module = __import__(module_path, fromlist=[class_name])
        return getattr(module, class_name)

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: type[T]) -> T: ...

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: None = None) -> Any: ...

    @staticmethod
    def _import_object(
        import_path: str, expected_type: type[T] | None = None
    ) -> T | Any:
        """Import object from module:object path.

        Args:
            import_path: Import path like "module.path:object_name"
            expected_type: Optional type to validate against

        Returns:
            Imported object (typed as expected_type if provided)

        Raises:
            ValueError: If path format is invalid
            ImportError: If module or object not found
            TypeError: If object doesn't match expected_type
        """
        if ":" not in import_path:
            raise ValueError(
                f"Invalid import path: {import_path}. "
                f"Expected format: 'module.path:object_name'"
            )

        module_path, object_name = import_path.split(":", 1)

        try:
            module = __import__(module_path, fromlist=[object_name])
            obj = getattr(module, object_name)
        except (ImportError, AttributeError) as e:
            raise ImportError(f"Failed to load object from {import_path}: {e}") from e

        if expected_type is not None and not isinstance(obj, expected_type):
            type_name = getattr(expected_type, "__name__", str(expected_type))
            raise TypeError(
                f"Object at {import_path} is not a {type_name} instance. "
                f"Got: {type(obj)}"
            )

        return obj
