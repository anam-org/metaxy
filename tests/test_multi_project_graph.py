"""Tests for multi-project feature graphs.

Tests that FeatureGraph can contain features from multiple projects and that
project information is properly tracked throughout the graph lifecycle.
"""

from __future__ import annotations

from syrupy.assertion import SnapshotAssertion

from metaxy import Feature, FeatureDep, FeatureKey, FeatureSpec, FieldKey, FieldSpec
from metaxy.config import MetaxyConfig
from metaxy.models.feature import FeatureGraph


def test_graph_contains_features_from_multiple_projects(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that a single FeatureGraph can contain features from different projects."""
    graph = FeatureGraph()

    # Create features from project_a
    config_a = MetaxyConfig(project="project_a")
    MetaxyConfig.set(config_a)

    with graph.use():

        class FeatureA1(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["project_a", "feature1"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        class FeatureA2(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["project_a", "feature2"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Create features from project_b in the same graph
    config_b = MetaxyConfig(project="project_b")
    MetaxyConfig.set(config_b)

    with graph.use():

        class FeatureB1(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["project_b", "feature1"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        class FeatureB2(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["project_b", "feature2"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Graph should contain all features
    assert len(graph.features_by_key) == 4

    # Verify projects are correct
    assert FeatureA1.project == "project_a"
    assert FeatureA2.project == "project_a"
    assert FeatureB1.project == "project_b"
    assert FeatureB2.project == "project_b"

    # Snapshot the graph structure
    projects_by_feature = {
        key.to_string(): feat.project for key, feat in graph.features_by_key.items()
    }

    assert projects_by_feature == snapshot

    MetaxyConfig.reset()


def test_graph_snapshot_includes_all_projects(snapshot: SnapshotAssertion) -> None:
    """Test that graph.to_snapshot() includes features from all projects."""
    graph = FeatureGraph()

    # Project A
    config_a = MetaxyConfig(project="snapshot_a")
    MetaxyConfig.set(config_a)

    with graph.use():

        class FeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_a"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Project B
    config_b = MetaxyConfig(project="snapshot_b")
    MetaxyConfig.set(config_b)

    with graph.use():

        class FeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_b"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Create snapshot
    snapshot_data = graph.to_snapshot()

    # Should have both features
    assert len(snapshot_data) == 2
    assert "feature_a" in snapshot_data
    assert "feature_b" in snapshot_data

    # Snapshot should track feature keys
    assert list(snapshot_data.keys()) == snapshot

    MetaxyConfig.reset()


def test_graph_snapshot_version_includes_all_projects(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that graph.snapshot_version is computed from all features regardless of project."""
    graph = FeatureGraph()

    # Add features from multiple projects
    config_a = MetaxyConfig(project="version_a")
    MetaxyConfig.set(config_a)

    with graph.use():

        class FeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_a"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    config_b = MetaxyConfig(project="version_b")
    MetaxyConfig.set(config_b)

    with graph.use():

        class FeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_b"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Get snapshot version
    snapshot_version = graph.snapshot_version

    # Should be deterministic
    assert len(snapshot_version) == 64  # SHA256 hex
    assert snapshot_version == snapshot

    # Verify it's based on all features
    # If we create a graph with only one feature, it should be different
    graph_single = FeatureGraph()

    with graph_single.use():

        class FeatureSingle(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_a"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    snapshot_version_single = graph_single.snapshot_version

    # Should be different (only one feature vs two)
    assert snapshot_version != snapshot_version_single

    MetaxyConfig.reset()


def test_graph_from_snapshot_preserves_projects(snapshot: SnapshotAssertion) -> None:
    """Test that FeatureGraph.to_snapshot() preserves project information."""
    original_graph = FeatureGraph()

    # Create features from multiple projects
    config_a = MetaxyConfig(project="restore_a")
    MetaxyConfig.set(config_a)

    with original_graph.use():

        class SnapshotFeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["restore", "feature_a"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    config_b = MetaxyConfig(project="restore_b")
    MetaxyConfig.set(config_b)

    with original_graph.use():

        class SnapshotFeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["restore", "feature_b"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Get original projects
    original_projects = {
        key.to_string(): feat.project
        for key, feat in original_graph.features_by_key.items()
    }

    # Create snapshot
    snapshot_data = original_graph.to_snapshot()

    # Verify that snapshot contains project information for all features
    for feature_key_str, feature_data in snapshot_data.items():
        assert "project" in feature_data
        assert "feature_tracking_version" in feature_data
        # Project should match what we expect
        if "feature_a" in feature_key_str:
            assert feature_data["project"] == "restore_a"
        elif "feature_b" in feature_key_str:
            assert feature_data["project"] == "restore_b"

    # Extract projects from snapshot
    snapshot_projects = {key: data["project"] for key, data in snapshot_data.items()}

    # Projects should be preserved in snapshot
    assert original_projects == snapshot_projects
    assert snapshot_projects == snapshot

    MetaxyConfig.reset()


def test_multi_project_dependency_graph(snapshot: SnapshotAssertion) -> None:
    """Test that features can depend on features from other projects."""
    graph = FeatureGraph()

    # Project A: upstream feature
    config_a = MetaxyConfig(project="upstream_project")
    MetaxyConfig.set(config_a)

    with graph.use():

        class UpstreamFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Project B: downstream feature depending on Project A's feature
    config_b = MetaxyConfig(project="downstream_project")
    MetaxyConfig.set(config_b)

    with graph.use():

        class DownstreamFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(key=FeatureKey(["upstream"]))],
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Both features should exist in graph
    assert len(graph.features_by_key) == 2

    # Verify projects
    assert UpstreamFeature.project == "upstream_project"
    assert DownstreamFeature.project == "downstream_project"

    # Verify dependency is tracked
    downstream_spec = graph.feature_specs_by_key[FeatureKey(["downstream"])]
    assert len(downstream_spec.deps or []) == 1
    assert downstream_spec.deps is not None  # Type assertion for basedpyright
    assert downstream_spec.deps[0].key == FeatureKey(["upstream"])

    # Snapshot the structure
    assert {
        "upstream_project": UpstreamFeature.project,
        "downstream_project": DownstreamFeature.project,
        "has_dependency": len(downstream_spec.deps or []) > 0,
    } == snapshot

    MetaxyConfig.reset()


def test_get_downstream_features_across_projects(snapshot: SnapshotAssertion) -> None:
    """Test that get_downstream_features works with multi-project graphs."""
    graph = FeatureGraph()

    # Root feature in project A
    config_a = MetaxyConfig(project="project_a")
    MetaxyConfig.set(config_a)

    with graph.use():

        class RootFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["root"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Mid-level feature in project B
    config_b = MetaxyConfig(project="project_b")
    MetaxyConfig.set(config_b)

    with graph.use():

        class MidFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["mid"]),
                deps=[FeatureDep(key=FeatureKey(["root"]))],
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Leaf feature in project C
    config_c = MetaxyConfig(project="project_c")
    MetaxyConfig.set(config_c)

    with graph.use():

        class LeafFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["leaf"]),
                deps=[FeatureDep(key=FeatureKey(["mid"]))],
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Get downstream features of root
    downstream = graph.get_downstream_features([FeatureKey(["root"])])

    # Should include both mid and leaf
    downstream_strings = [key.to_string() for key in downstream]

    assert "mid" in downstream_strings
    assert "leaf" in downstream_strings
    assert len(downstream) == 2

    # Verify projects
    projects = {
        "root": RootFeature.project,
        "mid": MidFeature.project,
        "leaf": LeafFeature.project,
        "downstream_keys": downstream_strings,
    }

    assert projects == snapshot

    MetaxyConfig.reset()


def test_graph_with_same_feature_key_different_projects() -> None:
    """Test that two graphs can have the same feature key in different projects.

    This is important for multi-tenant scenarios where different projects
    might use the same feature naming conventions.
    """
    # Graph 1: project_x
    graph_x = FeatureGraph()
    config_x = MetaxyConfig(project="project_x")
    MetaxyConfig.set(config_x)

    with graph_x.use():

        class FeatureInX(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["common", "feature"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Graph 2: project_y with SAME feature key
    graph_y = FeatureGraph()
    config_y = MetaxyConfig(project="project_y")
    MetaxyConfig.set(config_y)

    with graph_y.use():

        class FeatureInY(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["common", "feature"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Both should exist in their respective graphs
    assert len(graph_x.features_by_key) == 1
    assert len(graph_y.features_by_key) == 1

    # Both should have the same feature key
    assert FeatureInX.spec().key == FeatureInY.spec().key

    # But different projects
    assert FeatureInX.project == "project_x"
    assert FeatureInY.project == "project_y"
    assert FeatureInX.project != FeatureInY.project

    # And same feature_version (computational definition identical)
    assert FeatureInX.feature_version() == FeatureInY.feature_version()

    MetaxyConfig.reset()
