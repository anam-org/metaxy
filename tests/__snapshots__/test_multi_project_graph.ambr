# serializer version: 1
# name: test_get_downstream_features_across_projects
  dict({
    'downstream_keys': list([
      'mid',
      'leaf',
    ]),
    'leaf': 'project_c',
    'mid': 'project_b',
    'root': 'project_a',
  })
# ---
# name: test_graph_contains_features_from_multiple_projects
  dict({
    'project_a/feature1': 'project_a',
    'project_a/feature2': 'project_a',
    'project_b/feature1': 'project_b',
    'project_b/feature2': 'project_b',
  })
# ---
# name: test_graph_from_snapshot_preserves_projects
  dict({
    'restore/feature_a': 'restore_a',
    'restore/feature_b': 'restore_b',
  })
# ---
# name: test_graph_snapshot_includes_all_projects
  list([
    'feature_a',
    'feature_b',
  ])
# ---
# name: test_graph_snapshot_version_includes_all_projects
  'c4fab32026447b2afc7cae3c4c0e42044ea9a43a5d2c4454f00a6fb6587277e1'
# ---
# name: test_multi_project_dependency_graph
  dict({
    'downstream_project': 'downstream_project',
    'has_dependency': True,
    'upstream_project': 'upstream_project',
  })
# ---
